### DISCUSSION TITLE
Proposal to Improve Synchronization in message_hook Function for IKEV2 Processing

### DISCUSSION THREAD

**Discussion by liudf0716** [strongswan/strongswan#2032] [General]
Hello teams!

I have a question about the function `message_hook` in the `ha_ike.c` file. When processing IKEV2, the code currently syncs on requests, as shown in the following code snippet:

```c
// Existing code...
if (plain && ike_sa->get_version(ike_sa) == IKEV2)
{
	if (message->get_exchange_type(message) != IKE_SA_INIT &&
	    message->get_request(message))
       {	/* we sync on requests, but skip it on IKE_SA_INIT */
	   ha_message_t *m;

	    if (incoming)
	   {
		m = ha_message_create(HA_IKE_MID_RESPONDER);
	   }
	   else
	  {
		m = ha_message_create(HA_IKE_MID_INITIATOR);
	   }
	   m->add_attribute(m, HA_IKE_ID, ike_sa->get_id(ike_sa));
	   m->add_attribute(m, HA_MID, message->get_message_id(message) + 1);
	   this->socket->push(this->socket, m);
	   this->cache->cache(this->cache, ike_sa, m);
    }
    ...
} 
```

However, I propose that it should sync on responses instead. For example, when the initiator launches the rekey process, it sends a `create_child_sa` packet, followed by an `HA_IKE_MID_INITIATOR` message to the back, increasing the back's Message ID. If a handoff occurs during this time, causing the `create_child_sa` packet to be dropped and the responder not receiving it, when the back becomes the master, it sends a DPD with an increased Message ID to remote peers. The remote peers, finding the Message ID greater than theirs, do not respond, leading to retransmission exceeding 5 times and tunnel destruction.

If we change the code as follows, such a problem can be avoided:

```c
// Proposed code...
if (plain && ike_sa->get_version(ike_sa) == IKEV2)
{
    // Code for syncing on responses...
    if (message->get_exchange_type(message) != IKE_SA_INIT &&
	!message->get_request(message))
   {	/* we sync on responses, but skip it on IKE_SA_INIT */
		ha_message_t *m;

		if (incoming)
		{
			m = ha_message_create(HA_IKE_MID_INITIATOR);
		}
		else
		{
			m = ha_message_create(HA_IKE_MID_RESPONDER);
		}
		m->add_attribute(m, HA_IKE_ID, ike_sa->get_id(ike_sa));
		m->add_attribute(m, HA_MID, message->get_message_id(message) + 1);
		this->socket->push(this->socket, m);
		this->cache->cache(this->cache, ike_sa, m);
	}
}
```

If you agree with my suggestion, I'm willing to submit a pull request to implement this change. Thank you!

**Comment by tobiasbrunner** [comment]
Note that the _ha_ plugin is currently not intended to be used on initiators.

> If a handoff occurs during this time, causing the `create_child_sa` packet to be dropped and the responder not receiving it, when the back becomes the master, it sends a DPD with an increased Message ID to remote peers.

The problem is that we don't know if the responder received the CREATE_CHILD_SA request or not (the response could get dropped). So the node that takes over can't just send an INFORMATIONAL/DPD request using the same MID that the original node used to send the CREATE_CHILD_SA request.

**Reply by liudf0716** [reply]
In my test case, during handoff execution,  the Create_Child_SA packet was not received by the responder . Subsequently, the device became the master and initiated DPD. The peers then detected that the MID value was greater than their own.

During handoff, I suspect a delay in strongSwan's detection of the state change might be causing the Create_Child_SA packet to be dropped by the data plane. While strongSwan is transitioning to the 'back' state, the data plane may already be operating in that state due to faster processing. This mismatch could lead the data plane to inadvertently drop the packet before strongSwan is ready to send it.
